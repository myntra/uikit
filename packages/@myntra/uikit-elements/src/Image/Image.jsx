import React, { Component } from 'react'
import PropTypes from 'prop-types'

import IntersectionObserver from 'intersection-observer-polyfill'

import classnames from './Image.module.css'

export const OBSERVER = Symbol('Image Observer')

export function createObserver(root) {
  root = root || document.body
  if (root[OBSERVER]) return root[OBSERVER]

  let counter = 0
  const handlers = new WeakMap()
  const wrapper = (root[OBSERVER] = {
    handlers,
    observe(element, handler) {
      handlers.set(element, handler)
      observer.observe(element)
      counter++
    },
    unobserve(element) {
      observer.unobserve(element)
      handlers.delete(element)
      counter--

      if (counter === 0) {
        observer.disconnect()
        delete root[OBSERVER]
      }
    }
  })
  const observer = new IntersectionObserver(
    entries => {
      entries.forEach(entry => {
        if (handlers.has(entry.target)) {
          const fn = handlers.get(entry.target)
          fn(entry, observer)
        }
      })
    },
    { root: root === document.body ? null : root, rootMargin: '0px', threshold: 0 }
  )

  wrapper.raw = observer

  return wrapper
}

/**
  A component to lazy load images.
  @since 0.3.0
  @status EXPERIMENTAL
  @example
  <Image  src="https://picsum.photos/500/600" width={500} height={600} />
 */

class Image extends Component {
  static propTypes = {
    /** @private */
    className: PropTypes.string,
    /** Image source */
    src: PropTypes.string,
    /** Image height */
    height: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
    /** Image width */
    width: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
    /** Lazy loading image */
    lazyLoad: PropTypes.bool
  }

  static defaultProps = {
    lazyLoad: true
  }

  constructor(props) {
    super(props)

    this.state = {
      status: 'loading',
      isIntersecting: false
    }

    this.ref = React.createRef()
  }

  componentDidMount() {
    this.downloadImage()
    this.attachIntersectionObserver()
  }

  componentDidUpdate(prevProps) {
    if (prevProps.src !== this.props.src) {
      this.setState({ status: 'loading' })
      this.downloadImage()
    }
  }

  componentWillUnmount() {
    this.observer.unobserve(this.ref.current)
    delete this.observer
  }

  downloadImage() {
    const schedule = window.requestIdleCallback || (fn => setTimeout(fn, 1))

    schedule(
      /* istanbul ignore next: non-imperative code to preload image. */
      () => {
        const image = new window.Image()
        image.addEventListener('load', () => this.setState({ status: 'loaded' }), false)
        image.addEventListener('error', () => this.setState({ status: 'failed' }), false)
        image.src = this.props.src
      },
      { timeout: 50 }
    )
  }

  attachIntersectionObserver() {
    if (!this.props.lazyLoad) {
      this.setState({ isIntersecting: true })
    } else if (this.ref.current) {
      ;(this.observer = createObserver(this.ref.current.offsetParent)).observe(
        this.ref.current,
        this.handleIntersection
      )
    }
  }

  handleIntersection = entry => {
    if (entry.isIntersecting) {
      this.setState({ isIntersecting: true })
      this.observer.unobserve(this.ref.current)
    }
  }

  render() {
    const { className, lazyLoad, height, width, ...forwardedProps } = this.props
    const { status, isIntersecting } = this.state

    return (
      <div ref={this.ref} className={classnames('image-container', { 'bg-error': status === 'failed' })}>
        {isIntersecting && status === 'loaded' ? (
          <img
            {...forwardedProps}
            key="image"
            height={height}
            width={width}
            className={classnames(className, 'img', status)}
          />
        ) : (
          <img
            key="image"
            height={height}
            width={width}
            src={'data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw=='}
          />
        )}
      </div>
    )
  }
}

export default Image
