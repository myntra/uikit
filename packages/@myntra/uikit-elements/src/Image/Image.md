#### Example: Masonry Image List

```jsx render editor
const style = {'breakInside': 'avoid-column', 'listStyleType': 'none', 'marginBottom': '16px'}
;
<ul style={{'columnGap': '16px', 'columnCount': '5'}}>
    <li style={style}>
        <Image src="https://picsum.photos/200/200" width={200} height={200} />
    </li>
    <li style={style}>
        <Image src="https://picsum.photos/200/300" width={200} height={300} />
    </li>
    <li style={style}>
        <Image src="https://picsum.photos/200/100" width={200} height={100} />
    </li>
    <li style={style}>
        <Image src="https://picsum.photos/200/400" width={200} height={400} />
    </li>
    <li style={style}>
        <Image src="https://picsum.photos/400/300" width={400} height={300} />
    </li>
    <li style={style}>
        <Image src="https://picsum.photos/400/200" width={400} height={200} />
    </li>
    <li style={style}>
        <Image src="https://picsum.photos/200/200" width={200} height={200} />
    </li>
    <li style={style}>
        <Image src="https://picsum.photos/300/300" width={300} height={300} />
    </li>
    <li style={style}>
        <Image src="https://picsum.photos/300/500" />
    </li>
</ul>
```

#### Example: Grid Image List

```jsx render editor
    <>
        <Grid>
            <Grid.Column><Image src="https://picsum.photos/300/200?g1" height={200} width={300} /></Grid.Column>
            <Grid.Column><Image src="https://picsum.photos/300/200?g2" height={200} width={300} /></Grid.Column>
            <Grid.Column><Image src="https://picsum.photos/300/200?g3" height={200} width={300} /></Grid.Column>
            <Grid.Column><Image src="https://picsum.photos/300/200?g4" height={200} width={300} /></Grid.Column>
        </Grid>
        <Grid>
            <Grid.Column><Image src="https://picsum.photos/300/200?g6" height={200} width={300} /></Grid.Column>
            <Grid.Column><Image src="https://picsum.photos/300/200?g7" height={200} width={300} /></Grid.Column>
            <Grid.Column><Image src="https://picsum.photos/300/200?g8" height={200} width={300} /></Grid.Column>
            <Grid.Column><Image src="https://picsum.photos/300/200?g9" height={200} width={300} /></Grid.Column>
        </Grid>
    </>
```

#### Example: Error Handling

```jsx render editor
<Image src="somerandomurl.jpg" height={200}  />
```
