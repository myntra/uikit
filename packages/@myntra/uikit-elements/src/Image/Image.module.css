.img {
  display: block;
  outline: none;
  animation: fadein 350ms;
}

@keyframes fadein {
  from {
    opacity: 0;
  }

  to {
    opacity: 1;
  }
}

.loading {
  display: none;
}

.image-container {
  display: inline-block;
}

.bg-error {
  background: var(--color-info);
}
