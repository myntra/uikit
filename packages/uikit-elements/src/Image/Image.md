#### Example: Grid Image List

```jsx render editor
<>
  <Grid>
    <Grid.Column><Image src="https://picsum.photos/300/200?g1" height={200} width={300} /></Grid.Column>
    <Grid.Column><Image src="https://picsum.photos/300/200?g2" height={200} width={300} /></Grid.Column>
    <Grid.Column><Image src="https://picsum.photos/300/200?g3" height={200} width={300} /></Grid.Column>
    <Grid.Column><Image src="https://picsum.photos/300/200?g4" height={200} width={300} /></Grid.Column>
  </Grid>
  <Grid>
    <Grid.Column><Image src="https://picsum.photos/300/200?g6" height={200} width={300} /></Grid.Column>
    <Grid.Column><Image src="https://picsum.photos/300/200?g7" height={200} width={300} /></Grid.Column>
    <Grid.Column><Image src="https://picsum.photos/300/200?g8" height={200} width={300} /></Grid.Column>
    <Grid.Column><Image src="https://picsum.photos/300/200?g9" height={200} width={300} /></Grid.Column>
  </Grid>
</>
```

#### Example: Error Handling

```jsx render editor
<Image src="somerandomurl.jpg" height={200}  />
```
