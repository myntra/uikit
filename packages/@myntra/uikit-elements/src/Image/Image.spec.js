import React from 'react'
import { shallow, mount } from 'enzyme'
import Image, { OBSERVER } from './Image'

describe('Image', () => {
  it('renders', () => {
    const wrapper = shallow(
      <Image src="https://picsum.photos/300/200" alt={'example'} lazyLoad={false} width={'100px'} />
    )

    expect(wrapper).toMatchSnapshot()
  })

  it('lazy loads image', () => {
    const url = 'https://picsum.photos/300/200'
    const wrapper = shallow(<Image src={url} />)

    expect(wrapper.find('img').html()).not.toEqual(expect.stringContaining(url))

    wrapper.setState({ status: 'loaded', isIntersecting: true })

    expect(wrapper.find('img').html()).toEqual(expect.stringContaining(url))
  })

  it('schedules image download using requestIdleCallback or setTimeout', () => {
    const spy = jest.spyOn(window, 'setTimeout')
    mount(<Image src="https://picsum.photos/300/200" />)

    expect(spy).toHaveBeenCalled()
    spy.mockClear()
  })

  it('reschedules image download on src change', done => {
    const spy = jest.spyOn(window, 'setTimeout')
    const wrapper = mount(<Image src="https://picsum.photos/300/200" />)
    expect(spy).toHaveBeenCalledTimes(1)

    wrapper.setProps({ src: 'https://picsum.photos/300/300' }, () => {
      expect(spy).toHaveBeenCalledTimes(2)
      done()
    })
    spy.mockClear()
  })

  describe('IntersectionObserver', () => {
    it('attaches intersection observer on mount', () => {
      const wrapper = mount(<Image src="https://picsum.photos/300/200" />)
      const { observer, ref } = wrapper.instance()

      expect(observer).toBeTruthy()
      expect(observer.handlers.get(ref.current)).toBeTruthy()
    })

    it('releases intersection observer on unmount', () => {
      const wrapper = mount(<Image src="https://picsum.photos/300/200" />)
      const { observer, ref } = wrapper.instance()

      wrapper.unmount()

      expect(observer.handlers.get(ref.current)).toBeFalsy()
    })

    it('calls correct intersection handler', () => {
      const wrapper = mount(<Image src="https://picsum.photos/300/200" />)
      const instance = wrapper.instance()
      const { observer } = instance

      expect(wrapper.state('isIntersecting')).toBe(false)

      observer.raw.handler([{ target: null, isIntersecting: true }])
      expect(wrapper.state('isIntersecting')).toBe(false)

      observer.raw.handler([{ target: instance.ref.current, isIntersecting: false }])
      expect(wrapper.state('isIntersecting')).toBe(false)

      observer.raw.handler([{ target: instance.ref.current, isIntersecting: true }])
      expect(wrapper.state('isIntersecting')).toBe(true)
    })

    it('unobserves on intersection', () => {
      const wrapper = mount(<Image src="https://picsum.photos/300/200" />)
      const { observer } = wrapper.instance()

      const spy = jest.spyOn(observer, 'unobserve')

      wrapper.instance().handleIntersection({ isIntersecting: true })

      expect(spy).toHaveBeenCalled()
      spy.mockClear()
    })

    it('disconnects if no active image instance', () => {
      delete document.body[OBSERVER]

      const wrapper = mount(<Image src="https://picsum.photos/300/200" />)
      const { observer } = wrapper.instance()

      const spy = jest.spyOn(observer.raw, 'disconnect')

      mount(<Image src="https://picsum.photos/300/200" />).unmount()

      expect(spy).not.toHaveBeenCalled()

      wrapper.unmount()

      expect(spy).toHaveBeenCalled()
    })
  })
})
