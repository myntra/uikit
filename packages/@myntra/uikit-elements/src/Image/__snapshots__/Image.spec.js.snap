// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image renders 1`] = `
<div
  className="image-container"
>
  <img
    key="image"
    src="data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw=="
    width="100px"
  />
</div>
`;
